### DISCUSSION TITLE
Follow-up

### DISCUSSION THREAD

**Discussion by nozerorma** [nozerorma/decont#1] [Q&A]
So I like how things are going with the script, everything seems to be working as expected. I added the decontaminants url to the url file, I don't know if that's what we are meant to do. I've written until the index build section. @tdido

**Comment by tdido** [comment]
That's not really the best way to do it, since it would download together with the data. You can just pass the URL directly to the script in the code. Maybe use a variable for clarity.

**Comment by nozerorma** [comment]
I filtered the urls with grep so it doesn't download both things at the same time, and so I could modify the output path and pass the filt argument

```
echo -e "Downloading required files...\n"
mkdir -p data
# Download and extract required genomes
for url in $(grep 'https' data/urls | grep -v 'contaminants' | sort -u)
do
        bash scripts/download.sh $url data yes 2>> log/errors.log
done

url=$(grep 'contaminants' data/urls)
# Download, extract and filter decontaminants database
bash scripts/download.sh $url res yes filt 2>> log/errors.log
```

**Reply by tdido** [reply]
Fair enough, that's a reasonable solution. It's a bit coupled to that particular dataset though. If you were to implement this in a real pipeline you could turn `urls` into a sort of configuration file like:

```
data https://data1
data https://data2
data https://data3
contaminants https://cont
```

And then use grep and cut to get the relevant urls.

**Reply by nozerorma** [reply]
Great, thanks for the input, did it that way and it makes more sense

**Comment by nozerorma** [comment]
One last thing, for the cleanup script I took an interactive approach, would like to know what you think about it before I submit the script.

From pipeline.sh
```
# Run cleanup script at start
bash scripts/cleanup.sh 2>> log/errors.log
```
From download.sh
```
if grep OK *.tmp
        then
                echo -e "Download integrity verified\n"
                rm *.tmp
                cd ..
        else
                echo -e "Download integrity could not be verified. Aborting...\n"
                cd ..
                bash scripts/cleanup.sh Y
                echo -e "\n############ Pipeline failed at $(date +'%H:%M:%S') ##############\n"
                exit 1
        fi
```
Cleanup script
```
if [ $# -eq 1 ]                                                                        
then
        # Works mainly as rollback script when certain situations do not comply
        
        debris=$1

        if [[ "${debris}" = *[Yy]* ]]
        then
                echo -e "\nCleaning everything up...\n"
                find data log out res -mindepth 1 ! \( -name 'urls' -o -name '.gitkeep' -o -name '*.log' \) -exec rm -rf {} \;
                # Exclude gitkeeps, urls, scripts and logs 
        fi

# Works both as standalone cleanup script and interactive startup cleanup script

else
        echo -e "Would you like to remove any remaining files from previous runs? <Y>/<n>\n"
        read debris
        echo

        if [[ ${debris} = *[Yy]* ]]
        then
                # Cleanup selectively

                echo -e "Which files would you like to remove? <data> <log> <out> <res>\n"
                read cleandebris
                expectedinput="data log out res"
                echo

                if [[ "$expectedinput" == *"$cleandebris"* ]]
                then
                        echo -e "Cleaning up $cleandebris\n"
                        find $cleandebris -mindepth 1 ! \( -name 'urls' -o -name '.gitkeep' \) -exec rm -rf {} \; 2> log/cleanup_err.log
                else
                        echo -e "No cleanup performed\n"
                fi
        else
                echo -e "No cleanup performed\n"
        fi
fi
```

**Reply by tdido** [reply]
Looks good. There may be cases where the automated removal of files may not be recommended, since error messages may not be enough and whatever is left in the directory may let you understand the error.

**Reply by nozerorma** [reply]
So it would be better to remove the rollback function if the md5 sums aren't verified?

**Reply by tdido** [reply]
I wasn't thinking about it as an automated rollback functionality actually, but rather a convenience script that you run when needed so you don't need to clean stuff manually. But feel free to repurpose!
